packages/hub/routes/scheduled-payments.ts

```diff
@@ -0,0 +1,149 @@
+      id: shortUUID.uuid(),
+      senderSafeAddress: attrs['sender-safe-address'],
+      moduleAddress: attrs['module-address'],
+      tokenAddress: attrs['token-address'],
+      amount: attrs.amount,
+      payeeAddress: attrs['payee-address'],
+      executionGasEstimation: attrs['execution-gas-estimation'],
+      maxGasPrice: attrs['max-gas-price'],
+      feeFixedUsd: attrs['fee-fixed-usd'],
+      feePercentage: attrs['fee-percentage'],
+      salt: attrs['salt'],
+      payAt: new Date(attrs['pay-at']),
+      recurringDayOfMonth: attrs['recurring-day-of-month'],
+      recurringUntil: attrs['recurring-until'],
+      validForDays: attrs['valid-for-days'],
+      spHash: attrs['sp-hash'],
+      chainId: attrs['chain-id'],
+      signature: attrs['signature'],
+    };
+
+    if (params.recurringDayOfMonth != null) {
+      params.payAt = calculateNextPayAt(new Date(), params.recurringDayOfMonth);
+    }
+
+    let errors = this.scheduledPaymentValidator.validate(params);
+    let hasErrors = Object.values(errors).flatMap((i) => i).length > 0;
+    if (hasErrors) {
+      ctx.status = 422;
+      ctx.body = {
+        errors: serializeErrors(errors),
+      };
+    } else {
+      // TODO: prevent double save (was there the same scheduled payment (apart from the salt) already saved recently?)
+      let prisma = await this.prismaManager.getClient();
+      let scheduledPaymentModule = await this.cardpay.getSDK('ScheduledPaymentModule', this.web3.getInstance());
+
+      // We create the scheduled payment in the database as a first step, so that we can be sure
+      // that the scheduled payment *is persisted* in the database before we try to map it with the transaction.
+      // We don't want to get into a situation where on chain call is triggered but the scheduled payment
+      // wasn't persisted in the database for some reason.
+      let scheduledPayment = await prisma.scheduledPayment.create({
+        data: params,
+      });
+
+      let getTxnHash = new Promise<void>((resolve) => {
+        scheduledPaymentModule.schedulePayment(
+          params.senderSafeAddress,
+          params.moduleAddress,
+          params.tokenAddress,
+          params.spHash,
+          signatureToVRS(params.signature),
+          {
+            onTxnHash: async (txnHash) => {
+              await prisma.scheduledPayment.update({
+                where: {
+                  id: scheduledPayment.id,
+                },
+                data: {
+                  creationTransactionHash: txnHash,
+                },
+              });
+
+              // It might take several minutes for the transaction to be mined, so we wait for it in the background. The client
+              // can poll the record to see when it's mined (creation_block_number attribute will be set).
+              await this.workerClient.addJob('scheduled-payment-on-chain-creation-waiter', {
+                scheduledPaymentId: scheduledPayment.id,
+              });
```

Comment:
The profile purchase code uses a “job ticket” pattern for a slow/chain operation; the endpoint returns a newly-created ticket that the consumer can poll at `/api/job-tickets/:id` to track progress. Would it make sense to use that pattern here?

Reply:
I thought about it. I think it's easier to just query for the scheduled payments. If `creation_block_number` is present on the record, it will mean the job finished successfully. In this case that pattern sounds a bit roundabout